fecgov/fec-cms
# Election-reporting dates tables 2.0

## Summary
- Resolves #5317
- Refined Wagtail template and spreadheet process for creating reporting dates tables like:
     - https://www.fec.gov/help-candidates-and-committees/dates-and-deadlines/2020-reporting-dates/congressional-pre-election-reporting-dates-2020/
     - https://www.fec.gov/help-candidates-and-committees/dates-and-deadlines/2020-reporting-dates/24-and-48-hour-reports-independent-expenditures-periods-congressional-primaries-2020/
- This  addresses the  completion criteria : "Update code/wagtail to allow for easier (and less chance for error) updates to the footnotes in large tables"
     - **No more editing HTML**: Switches to using  table-block field instead of raw html block. The  editor can add  Footnotes/Header-notes  to any cell in the table  with a tilde(~) character. To both the table header cells and the table body cells. 
     - **No more editing JSON**: The master Footnotes/Header-notes lists are added in Wagtail rich text fields, making it easier to re-order or edit the  description.
- **Safely escaped JSON**: Using Django's  [json_script](https://docs.djangoproject.com/en/4.2/ref/templates/builtins/#json-script)  filter,   data pulled from the database into the page for  use by JS is now fully escaped, mitigating XSS vulnerability.
- **Simplified spreadsheet** The spreadsheet used for initial setup is much simpler, allowing the editor to copy `columns/rows in Excel` to `columns/rows in Wagtail Tableblock` (no more giant blob of html to copy/paste or sift through to edit)
    - The Example 2024 spreadsheet (with data and html tabs) used to populate the Wagtail table block is https://docs.google.com/spreadsheets/d/1gvJ1_Dzdk7dbrIgwGZDgVPVABlCOOElT/edit#gid=538441488
    
**Note:** Past year's reporting dates tables use the old `repoting-dates-tables.js` so until they are archived, that script must remain in codebase.

Related: 
#4408
#4406
#3817

### Required reviewers

One frontend, one content

## Impacted areas of the application:
This creates a new Wagtail template model with migrations. Also add a new JS file. Until all previous-year reporting dates pages are archived(the ones that use full-width template) the legacy JS needs to remain in codebase)

**Modified:**
fec/static/css/customize-editor.css
fec/static/scss/components/_table-styles.scss
fec/home/templatetags/filters.py
fec/home/models.py
fec/home/blocks.py

**New:**
fec/static/js/pages/election-reporting-dates-tables.js
fec/home/templates/blocks/reporting-dates-table-block.html
fec/home/templates/home/reporting_dates_table.html
fec/home/migrations/0129_reportingdatestable.py

**Temporarily modified:**
tasks.py

## Related PRs

## How to test 
Pushed to dev for testing. [Example page](https://dev.fec.gov/help-candidates-and-committees/dates-and-deadlines/2023-reporting-dates/test-cong-pre-elec-2/): 
- checkout branch
- `./manage,py migrate`
- `npm run build`
-  Create a new `ReportingDatesTables` page in Wagtail
- Copy-paste cells from the [example spreadsheet](https://docs.google.com/spreadsheets/d/1gvJ1_Dzdk7dbrIgwGZDgVPVABlCOOElT/edit#gid=538441488) Cong-pre-elec NEW tab into a `Dates table` field.
- Create `Header notes` and `Footnotes` sections in the template
- Preview or publish to view table and test footnote renderings. Test Header notes popup dialogue
- Test adding more footnotes to table dates with tilde (`~`) character
- Test state filter dropdown and confirm zebra striping is correct on table
- [ ] **Revert `tasks.py` before merge**

## System architecture updates (if applicable)

(If this pull request changes our [current system diagram](https://github.com/fecgov/FEC/wiki/2.-FEC-system-diagram), include a description of those changes here and create a new ticket to update the system diagram)
